shared/src/main/scala/amf/core/client/scala/model/domain/Graph.scala

```diff
@@ -7,60 +7,64 @@ import amf.core.internal.parser.domain.{Annotations, FieldEntry, Value}
 
 case class Graph(e: DomainElement) {
 
-  def removeField(uri: String): this.type = {
+  private[amf] def removeField(uri: String): this.type = {
     e.fields.remove(uri)
     this
   }
 
+  private[amf] def scalarByField(field: Field): Seq[Any] = scalarByProperty(field.value.iri())
+
+  private[amf] def containsField(f: Field): Boolean = properties().contains(f.toString)
+
+  private[amf] def annotationsForValue(f: Field): Annotations =
+    e.fields.getValueAsOption(f).map(_.annotations).getOrElse(Annotations())
+
+  private[amf] def patchField(patchField: Field, patchValue: Value): Unit = {
+    e.set(patchField, patchValue.value, annotationsForValue(patchField))
+  }
+
+  private[amf] def patchObj(patchField: Field, mergedNode: AmfObject): Unit = {
+    e.set(patchField, mergedNode, annotationsForValue(patchField))
+  }
+
+  private[amf] def patchSeqField(patchField: Field, objs: Seq[AmfElement]): Unit = {
+    e.set(patchField, AmfArray(objs), annotationsForValue(patchField))
+  }
+
   def types(): Seq[String] = e.meta.`type`.map(_.iri()).distinct
 
   def properties(): Seq[String] = e.fields.fields().map(_.field.value.iri()).toSeq
 
-  def scalarByField(field: Field): Seq[Any] = scalarByProperty(field.value.iri())
+  def containsProperty(uri: String): Boolean = {
+    properties().contains(Namespace.defaultAliases.uri(uri).iri())
+  }
 
-  def scalarByProperty(propertyId: String): Seq[Any] = {
+  def getObjectByProperty(uri: String): Seq[DomainElement] = {
     e.fields.fields().find { f: FieldEntry =>
-      f.field.value.iri() == Namespace.defaultAliases.uri(propertyId).iri()
+      f.field.value.iri() == Namespace.defaultAliases.uri(uri).iri()
     } match {
       case Some(fieldEntry) =>
         fieldEntry.element match {
-          case scalar: AmfScalar                    => List(scalar.value)
-          case arr: AmfArray if arr.values.nonEmpty => arr.values.toList
-          case _                                    => List()
+          case entity: DomainElement => List(entity)
+          case arr: AmfArray if arr.values.nonEmpty && arr.values.head.isInstanceOf[DomainElement] =>
+            arr.values.map(_.asInstanceOf[DomainElement]).toList
+          case _ => List()
         }
       case None => List()
     }
   }
 
-  def getObjectByPropertyId(propertyId: String): Seq[DomainElement] = {
+  def scalarByProperty(uri: String): Seq[Any] = {
```

Comment:
We could return `AMFElement` here as it is in the `amf.core.client.scala` package

Reply:
AMFElement is just an interace for any kind of value. This method search specify for Scalars, so we could return an AmfScalar, but the idea is to get the literal value.
Remember that Graph is a client interface, all client Graphs should be equivalent